### PR TITLE
{Network} `az network application-gateway private-link`: Migrate commands to codegen v2

### PR DESCRIPTION
**Pull request by necusjz**
**Related command**

az network application-gateway private-link

**Description**

Total: 10 commands

**Testing Guide**

Evidences for simplifying the logic of `private-link add`:
1. When we input a non-existing subnet ID, service will raise:
> Code: InvalidResourceReference
Message: Resource /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/ethan-rg/providers/Microsoft.Network/virtualNetworks/ag-13Vnet/subnets/default2 referenced by resource /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/ethan-rg/providers/Microsoft.Network/applicationGateways/ag-13 was not found. Please make sure that the referenced resource exists, and that both resources are in the same region.

Hence, when there is no subnet like that, we can just create a new subnet to simplify the logic.

2. When we create a new subnet with an existing prefix, service will raise:
> Code: NetcfgInvalidSubnet 
Message: Subnet 'default2' is not valid in virtual network 'ag-13Vnet'.

Hence, we don't need to put such error handling logic in the codes.

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Network

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 1]
```diff
@@ -6734,8 +6900,10 @@ def delegation_trans(index, service_name):
             element_transformer=lambda _, policy_id: {"id": policy_id}
         )
         # use string instead of bool
-        args.private_endpoint_network_policies = args.disable_private_endpoint_network_policies
-        args.private_link_service_network_policies = args.disable_private_link_service_network_policies
+        if has_value(args.disable_private_endpoint_network_policies):
+            args.private_endpoint_network_policies = args.disable_private_endpoint_network_policies
+        if has_value(args.disable_private_link_service_network_policies):
+            args.private_link_service_network_policies = args.disable_private_link_service_network_policies
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:6906]: why add has_value verification here?

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 2]
```diff
@@ -703,6 +706,163 @@ def pre_instance_update(self, instance):
...
+        )
+        for fic in instance.properties.frontend_ip_configurations:
+            if hasattr(fic.properties, "private_link_configuration") \
+                    and fic.properties.private_link_configuration.id == private_link_id:
+                err_msg = "Frontend IP already reference an existing private link."
+                raise ValidationError(err_msg)
+        # associate private link with frontend IP configuration
+        for fic in instance.properties.frontend_ip_configurations:
+            if fic.name == args.frontend_ip:
+                setattr(fic.properties, "private_link_configuration", {"id": private_link_id})
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:767]: why use setattr here?

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 3]
```diff
@@ -703,6 +706,163 @@ def pre_instance_update(self, instance):
...
+            subscription=self.ctx.subscription_id,
+            resource_group=args.resource_group,
+            namespace="Microsoft.Network",
+            type="applicationGateways",
+            name=args.gateway_name,
+            child_type_1="privateLinkConfigurations",
+            child_name_1=args.name
+        )
+        for fic in instance.properties.frontend_ip_configurations:
+            if hasattr(fic.properties, "private_link_configuration") \
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:760]: why need to use hasattr? Is fic polymorphism?

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 4]
```diff
@@ -703,6 +706,163 @@ def pre_instance_update(self, instance):
...
+    def pre_instance_delete(self):
+        args = self.ctx.args
+        instance = self.ctx.vars.instance
+        for plc in instance.properties.private_link_configurations:
+            if plc.name == args.name:
+                to_be_removed = plc
+                break
+        else:
+            err_msg = "Private link doesn't exist."
+            raise ValidationError(err_msg)
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:834]: Does old version of python version support `for...else...`?

File: src/azure-cli/azure/cli/command_modules/network/custom.py (hunk 3 above)
Comment (necusjz) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:760]: will double check

File: src/azure-cli/azure/cli/command_modules/network/custom.py (hunk 4 above)
Comment (necusjz) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:834]: It has been introduced from 1990...

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 5]
```diff
@@ -703,6 +706,163 @@ def pre_instance_update(self, instance):
...
+            resource_group=args.resource_group,
+            namespace="Microsoft.Network",
+            type="applicationGateways",
+            name=args.gateway_name,
+            child_type_1="privateLinkConfigurations",
+            child_name_1=args.name
+        )
+        for fic in instance.properties.frontend_ip_configurations:
+            if has_value(fic.properties.private_link_configuration) \
+                    and fic.properties.private_link_configuration.id == private_link_id:
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:761]: I think the following expression can also work.
```suggestion
            if fic.properties.private_link_configuration.id == private_link_id:
```

Comment (kairu-ms) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:761]: But it's good to add has_value check

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 6]
```diff
@@ -703,6 +706,163 @@ def pre_instance_update(self, instance):
...
+                if plc.name == args.name:
+                    err_msg = "Private link name duplicates."
+                    raise ValidationError(err_msg)
+        # prepare subnet for new private link
+        from .aaz.latest.network.vnet import Show
+        rid = instance.properties.gateway_ip_configurations[0].properties.subnet.id.to_serialized_data()
+        vnet_name = parse_resource_id(rid)["name"]
+        vnet = Show(cli_ctx=self.cli_ctx)(command_args={
+            "name": vnet_name,
+            "resource_group": args.resource_group
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:780]: Will the resource_group of rid be different from args.resource_group?

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 7]
```diff
@@ -703,6 +706,163 @@ def pre_instance_update(self, instance):
...
+        })
+        for subnet in vnet["subnets"]:
+            if subnet["name"] == args.subnet:
+                err_msg = "Subnet name duplicates. In order to use existing subnet, please enter subnet ID."
+                raise ValidationError(err_msg)
+            cond1 = subnet["addressPrefix"] == args.subnet_prefix
+            cond2 = "addressPrefixes" in subnet and args.subnet_prefix in subnet["addressPrefixes"]
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:787]: 'addressPrefix' and 'addressPrefixes' may be not exist in subnet.

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 8]
```diff
@@ -703,6 +706,163 @@ def pre_instance_update(self, instance):
...
+        rid = instance.properties.gateway_ip_configurations[0].properties.subnet.id.to_serialized_data()
+        vnet_name = parse_resource_id(rid)["name"]
+        vnet = Show(cli_ctx=self.cli_ctx)(command_args={
+            "name": vnet_name,
+            "resource_group": args.resource_group
+        })
+        for subnet in vnet["subnets"]:
+            if subnet["name"] == args.subnet:
+                err_msg = "Subnet name duplicates. In order to use existing subnet, please enter subnet ID."
+                raise ValidationError(err_msg)
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:785]: My concern is that, if the subnet["name"] == args.subnet, why not replase args.subnet by subnet["id"]?

**Review by kairu-ms — APPROVED** [review]

**Pull request merged**